libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,383 @@
+package keystore
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"golang.org/x/crypto/pbkdf2"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/file"
+)
+
+const (
+	filePermission = 0600
+
+	// Encryption Related constants
+	iVLength        = 12
+	saltLength      = 64
+	iterationsCount = 10000
+	keyLength       = 32
+)
+
+// Version of the keystore format, will be added at the beginning of the file.
+var version = []byte("v1")
+
+// FileKeystore Allows to store key / secrets pair securely into an encrypted local file.
+type FileKeystore struct {
+	sync.RWMutex
+	Path     string
+	secrets  map[string]serializableSecureString
+	dirty    bool
+	password *SecureString
+}
+
+// Allow the original SecureString type to be correctly serialized to json.
+type serializableSecureString struct {
+	*SecureString
+	Value []byte `json:"value"`
+}
+
+// NewFileKeystore returns an new File based keystore or an error, currently users cannot set their
+// own password on the keystore, the default password will be an empty string. When the keystore
+// is initialied the secrets are automatically loaded into memory.
+func NewFileKeystore(keystoreFile string) (Keystore, error) {
```

Comment:
As there are potentially coming more key stores in the future, I'm thinking if we should put each in it's own package. `keystore/file/...` for this one, so this would be just `New`. BTW that is refactoring that could also be done in a second step.

Reply:
Agree with that, since the pluggable part of the keystore is not complete, I think we should it when we add more.